src/printer.rs

```diff
@@ -43,43 +42,48 @@ impl Printer {
         }
     }
 
-    fn print_json(&mut self, text: &str) -> io::Result<()> {
-        // This code is a little thorny because of ownership issues.
-        // We have to keep the indented text alive until the end of the function.
-        let indent_result = match self.indent_json {
-            true => Some(indent_json(text)),
-            false => None,
-        };
-        let text = match &indent_result {
-            Some(Ok(result)) => result.as_str(),
-            _ => text,
-        };
-        if self.color {
-            colorize(text, "json", &self.theme, &mut self.buffer)
-        } else {
-            self.buffer.print(text)
-        }
+    fn colorize(&mut self, text: &str, syntax: &'static str) -> io::Result<()> {
+        colorize(text, syntax, &self.theme, &mut self.buffer)
     }
 
-    fn print_xml(&mut self, text: &str) -> io::Result<()> {
-        if self.color {
-            colorize(text, "xml", &self.theme, &mut self.buffer)
-        } else {
-            self.buffer.print(text)
+    fn dump(&mut self, reader: &mut impl Read) -> io::Result<()> {
+        io::copy(reader, &mut self.buffer)?;
+        Ok(())
+    }
+
+    fn print_json(&mut self, body: &mut impl Read) -> io::Result<()> {
+        match (self.indent_json, self.color) {
+            (false, false) => self.dump(body),
+            (true, false) => get_json_formatter().format_stream(body, &mut self.buffer),
+            (false, true) => {
+                let mut buf = Vec::new();
+                body.read_to_end(&mut buf)?;
+                let text = String::from_utf8_lossy(&buf);
+                self.colorize(&text, "json")
+            }
+            (true, true) => {
+                let mut buf = Vec::new();
+                get_json_formatter().format_stream(body, &mut buf)?;
+                let text = String::from_utf8_lossy(&buf);
+                self.colorize(&text, "json")
+            }
         }
     }
 
-    fn print_html(&mut self, text: &str) -> io::Result<()> {
+    fn print_syntax(&mut self, body: &mut impl Read, syntax: &'static str) -> io::Result<()> {
         if self.color {
-            colorize(text, "html", &self.theme, &mut self.buffer)
+            let mut buf = Vec::new();
+            body.read_to_end(&mut buf)?;
+            let text = String::from_utf8_lossy(&buf);
+            self.colorize(&text, syntax)
```

Comment:
Leaving a note to myself:

This (as well as `print_json()`'s coloring) is wrong, because it doesn't pay attention to `self.stream`.

A [`BufReader`](https://doc.rust-lang.org/std/io/struct.BufReader.html) could be used to read [line by line](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines). That's better than manually calling `read()`, which could split at awkward places, even inside a UTF-8 character. Using `read_until()` might be best so the final newline isn't lost.

Reply:
[`LineWriter`](https://doc.rust-lang.org/std/io/struct.LineWriter.html) turned out to be easier to hook up to `jsonxf` and somehow performs much better.